### PR TITLE
Refactor errors 

### LINKED ISSUE #150: Failure crate: "soft ecosystem deprecated"
- Related discussion: https://github.com/interchainio/tendermint-rs/pull/149#issuecomment-584448738

- Reasoning in abscissa repo: https://github.com/iqlusioninc/abscissa/issues/144

- Good read: https://blog.yoshuawuyts.com/error-handling-survey/

Currently:
https://github.com/interchainio/tendermint-rs/blob/f9287e04ee84e6659ea209529a1bb700ab53e45b/tendermint/Cargo.toml#L36

Used in:
 - [ ] tendermint-rs/error.rs 
 - [ ] tendermint-rs/amino_types/validate.rs
 - [ ] tendermint-rs/rpc/error.rs

We should switch to a combination of thiserror + anonmaly for the whole code base as done in #151 and #149 for the light client module.

### PR DESCRIPTION
**Pull request by Liamsi**
Consistent error handling across the repo. 
I have written up some notes briefly explaining what "consistent" means here. ~~Will post them~~ Posted in https://github.com/interchainio/tendermint-rs/issues/150#issuecomment-586148337. They can be used for an ADR if that's desired.

closes #150 

* [ ] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGES.md

### CODE REVIEW

**Comment by Liamsi** [comment]
codecov wasn't reporting because of a missing report for the base commit. Deleted its comment.

File: tendermint/src/amino_types/block_id.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-use super::validate::{ConsensusMessage, ValidationError, ValidationErrorKind::*};
+use super::validate::{ConsensusMessage, Kind::*};
```
Comment (xla) [inline tendermint/src/amino_types/block_id.rs]: Why are we sticking with the wild-card import, can we qualify what we exactly need?

File: tendermint/src/amino_types/proposal.rs [hunk 2]
```diff
@@ -4,8 +4,9 @@ use super::{
     remote_error::RemoteError,
     signature::{SignableMsg, SignedMsgType},
     time::TimeMsg,
-    validate::{ConsensusMessage, ValidationError, ValidationErrorKind::*},
+    validate::{ConsensusMessage, Kind::*},
 };
+use crate::amino_types::validate;
```
Comment (xla) [inline tendermint/src/amino_types/proposal.rs]: Why not adding a `self` to the imports above?

File: tendermint/src/amino_types/proposal.rs [hunk 3]
```diff
@@ -4,8 +4,9 @@ use super::{
     remote_error::RemoteError,
     signature::{SignableMsg, SignedMsgType},
     time::TimeMsg,
-    validate::{ConsensusMessage, ValidationError, ValidationErrorKind::*},
+    validate::{ConsensusMessage, Kind::*},
```
Comment (xla) [inline tendermint/src/amino_types/proposal.rs]: Wild-card import.

File: tendermint/src/amino_types/proposal.rs [hunk 4]
```diff
@@ -172,7 +173,7 @@ impl SignableMsg for SignProposalRequest {
 }
 
 impl ConsensusMessage for Proposal {
-    fn validate_basic(&self) -> Result<(), ValidationError> {
+    fn validate_basic(&self) -> Result<(), Error> {
```
Comment (xla) [inline tendermint/src/amino_types/proposal.rs]: As there are multiple error types used in this module, what about qualifying all of them. This one `errror::Error` and `validate::Error` for the others.

File: tendermint/src/amino_types/signature.rs [hunk 5]
```diff
@@ -15,7 +15,7 @@ pub trait SignableMsg {
 
     /// Set the Ed25519 signature on the underlying message
     fn set_signature(&mut self, sig: &ed25519::Signature);
-    fn validate(&self) -> Result<(), ValidationError>;
+    fn validate(&self) -> Result<(), Error>;
```
Comment (xla) [inline tendermint/src/amino_types/signature.rs]:
```suggestion
    fn validate(&self) -> Result<(), validate::Error>;
```

**Comment by Liamsi** [comment]
Merging as is now. Will address below in a separate PR which will add an ADR, too.

> Generally I'd err on the side of more qualification so use crate::error and then error::Error and error::Kind.

**Pull request merged**